src/varlinkctl/varlinkctl.c

```diff
@@ -289,56 +289,93 @@ typedef struct GetInterfaceDescriptionData {
+        _cleanup_strv_free_ char **auto_interfaces = NULL;
+        char **interfaces;
+        const char *url;
         int r;
 
-        assert(argc == 3);
+        assert(argc >= 2);
         url = argv[1];
-        interface = argv[2];
+        interfaces = strv_skip(argv, 2);
 
         r = varlink_connect_auto(&vl, url);
         if (r < 0)
                 return r;
 
-        JsonVariant *reply = NULL;
-        r = varlink_callb_and_log(
-                        vl,
-                        "org.varlink.service.GetInterfaceDescription",
-                        &reply,
-                        JSON_BUILD_OBJECT(JSON_BUILD_PAIR_STRING("interface", interface)));
-        if (r < 0)
-                return r;
+        if (strv_isempty(interfaces)) {
+                JsonVariant *reply = NULL;
 
-        pager_open(arg_pager_flags);
+                /* If no interface is specified, introspect all of them */
 
-        if (FLAGS_SET(arg_json_format_flags, JSON_FORMAT_OFF)) {
-                static const struct JsonDispatch dispatch_table[] = {
-                        { "description",  JSON_VARIANT_STRING, json_dispatch_const_string, 0, JSON_MANDATORY },
+                r = varlink_call_and_log(vl, "org.varlink.service.GetInfo", /* parameters= */ NULL, &reply);
+                if (r < 0)
+                        return r;
+
+                const struct JsonDispatch dispatch_table[] = {
+                        { "interfaces", JSON_VARIANT_ARRAY, json_dispatch_strv, PTR_TO_SIZE(&auto_interfaces), JSON_MANDATORY },
                         {}
                 };
-                _cleanup_(varlink_interface_freep) VarlinkInterface *vi = NULL;
-                const char *description = NULL;
-                unsigned line = 0, column = 0;
 
-                r = json_dispatch(reply, dispatch_table, JSON_LOG, &description);
+                r = json_dispatch(reply, dispatch_table, JSON_LOG|JSON_ALLOW_EXTENSIONS, NULL);
                 if (r < 0)
                         return r;
 
-                /* Try to parse the returned description, so that we can add syntax highlighting */
-                r = varlink_idl_parse(ASSERT_PTR(description), &line, &column, &vi);
-                if (r < 0) {
-                        log_warning_errno(r, "Failed to parse returned interface description at %u:%u, showing raw interface description: %m", line, column);
+                if (strv_isempty(auto_interfaces))
+                        return log_error_errno(SYNTHETIC_ERRNO(ENXIO), "Service doesn't report any implement interfaces.");
 
-                        fputs(description, stdout);
-                        if (!endswith(description, "\n"))
-                                fputs("\n", stdout);
-                } else {
-                        r = varlink_idl_dump(stdout, /* use_colors= */ -1, vi);
+                strv_sort(auto_interfaces);
+                strv_uniq(auto_interfaces);
+                interfaces = auto_interfaces;
```

Comment:
Not a matter, but these three lines can be unified:
`interfaces = strv_uniq(strv_sort(auto_interfaces));`

Reply:
(BTW, I'd like to find unique entries first before sorting)

Reply:
> (BTW, I'd like to find unique entries first before sorting)

Maybe, we should introduce `strv_sort_uniq()` like `sort -u`. Then, we can assume the strv is sorted, then uniq can become faster.

Reply:
yeah, that would make sense, but let's do that another time.